### PR TITLE
FLE Guide - Requirements section (DOCSP-6419)

### PR DESCRIPTION
**Pull request by ccho-mongodb**
**JIRA:**
https://jira.mongodb.org/browse/DOCSP-6419

**Staging:**
https://docs-mongodbcom-staging.corp.mongodb.com/ecosystem/docsworker/fle-guide-reqs/use-cases/sensitive-data-encryption.html

### CODE REVIEW

File: source/use-cases/sensitive-data-encryption.txt [hunk 1]
```diff
@@ -40,9 +40,42 @@ Procedure
...
 
-- Point to setup and install directions
-- Nullam imperdiet lorem vitae vulputate lacinia.
-- Donec eget velit tincidunt, gravida diam ac, efficitur lacus.
+MongoDB Server 4.2 Enterprise
+  * For installation instructions, refer to the 
+    `Enterprise Edition Installation Tutorials <https://docs.mongodb.com/manual/installation/#mongodb-enterprise-edition-installation-tutorials>`_.
+
+
+MongoDB Driver Compatible with CSFLE
```
[the remark this answers is not in the document]
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: We introduce and define CSFLE in the intro section, it's just not included in this PR. https://github.com/mongodb/docs-ecosystem/pull/526/files#diff-c5c354e57498148e8ef37f38712ce945R25

File: source/use-cases/sensitive-data-encryption.txt [hunk 2]
```diff
@@ -40,9 +40,42 @@ Procedure
...
+
+
+AWS Key Management Service (KMS)
+  * AWS KMS is required for the secure deployment of CSFLE.
+
+  * Compatible drivers currently include integration with AWS KMS. Support for
+    other KMS providers may be added in the future.
+
+  * The client application will need AWS KMS permissions to the GenerateDataKey,
+    Encrypt, and Decrypt actions.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: I think we can focus this a bit more on the _what_ rather than _why_. Comments for each AWS KMS bullet:

- "AWS KMS is required for the secure deployment of CSFLE."
  - We can remove the first bullet since we explain the _why_ of a requirement in the guide. 

- "Compatible drivers currently include integration with AWS KMS. Support for other KMS providers may be added in the future."
  - "Compatible _MongoDB_ drivers currently support key management through AWS KMS. Support for other KMS providers may be added in the future."

- "The client application will need AWS KMS permissions to the GenerateDataKey, Encrypt, and Decrypt actions."
  - I believe these are, specifically, "IAM role permissions" so let's call out that they need an IAM user with these permissions
  - Let's move this to be the first bullet point since it's the only actual requirement here.
  - nit: thoughts on monospacing the permission names? e.g. ``` ``GenerateDataKey`` ```

File: source/use-cases/sensitive-data-encryption.txt [hunk 3]
```diff
@@ -40,9 +40,42 @@ Procedure
...
+
+MongoDB Authorization Privileges
+  * In order to create data encryption keys, the client needs the
+    `ReadWrite <https://docs.mongodb.com/manual/reference/built-in-roles/#readWrite>`_ 
+    built-in role on the key vault collection which is specified in the client 
+    configuration.
+
+  * If no additional data encryption keys need to be created, the client can 
+    be downgraded to the `listCollections <https://docs.mongodb.com/manual/reference/privilege-actions/#listCollections>`_
+    privilege on the key vault collection.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: nit: Let's not frame this as "downgraded" since they likely wouldn't ever have the elevated privileges in the first place

File: source/use-cases/sensitive-data-encryption.txt [hunk 4]
```diff
@@ -40,9 +40,42 @@ Procedure
...
+
+
+Server or VM Permissions
+  * The client application or a privileged user needs permissions to start the 
+    `mongocryptd <https://docs.mongodb.com/manual/core/security-client-side-encryption/#field-level-encryption-mongocryptd>`_
+    process on the host. While the drivers contain code to spawn the process.
```
Comment (nlarew) [inline source/use-cases/sensitive-data-encryption.txt]: > While the drivers contain code to spawn the process.

Looks like an incomplete fragment slipped in here

File: source/use-cases/sensitive-data-encryption.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: I can probably reword this to "Compatible MongoDB Driver" if we want to avoid using any abbreviations, but that is likely less clear. I would like to avoid any conceptual explanations for a "Requirements" section, but am happy to change it to whatever works best.

File: source/use-cases/sensitive-data-encryption.txt (hunk 4 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Whoops, will remove.

File: source/use-cases/sensitive-data-encryption.txt [hunk 5]
```diff
@@ -40,9 +40,42 @@ Procedure
...
+    `Enterprise Edition Installation Tutorials <https://docs.mongodb.com/manual/installation/#mongodb-enterprise-edition-installation-tutorials>`_.
+
+
+MongoDB Driver Compatible with CSFLE 
+  * For a list of drivers that support CSFLE, refer to the
+    `driver compatibility table <https://docs.mongodb.com/manual/core/security-client-side-encryption/#driver-compatibility-table>`_.
+
+
+AWS Key Management Service (KMS)
+  * AWS KMS is required for the secure deployment of CSFLE.
```
[the remark this answers is not in the document]
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks for the link! I will reword this to match the article content.

File: source/use-cases/sensitive-data-encryption.txt (hunk 2 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Thanks for the suggestions, will incorporate the first two.

I will omit the third bullet point and rely on the link Moe posted (https://docs.atlas.mongodb.com/security-aws-kms/) to explain the setup including permissions. This might be a hint that there should be guides that instruct both DBA/DeevOps and Developers to do an end-to-end setup (if AWS KMS is used).

It does not bother me to call these IAM role permissions, but I think at least Encrypt/Decrypt are "key policies" that are defined on specific customer master keys that are then added to an IAM role. Pretty confusing. See: https://docs.aws.amazon.com/kms/latest/developerguide/kms-api-permissions-reference.html

File: source/use-cases/sensitive-data-encryption.txt [hunk 6]
```diff
@@ -40,9 +40,42 @@ Procedure
...
+    `ReadWrite <https://docs.mongodb.com/manual/reference/built-in-roles/#readWrite>`_ 
+    built-in role on the key vault collection which is specified in the client 
+    configuration.
+
+  * If no additional data encryption keys need to be created, the client can 
+    be downgraded to the `listCollections <https://docs.mongodb.com/manual/reference/privilege-actions/#listCollections>`_
+    privilege on the key vault collection.
+
+
+Server or VM Permissions
```
Comment (ccho-mongodb) [inline source/use-cases/sensitive-data-encryption.txt]: Uncertain whether this is conventional terminology as well as whether it is clear. The next closest term I could think of is Client-Host. Any suggestions?

File: source/use-cases/sensitive-data-encryption.txt (hunk 3 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: I was trying to identify the difference between what they permissions they would need for the data key creation step and everything following that.

On second thought, this auth requirement might be trivial to even include in this guide, especially if we assume the user is running through the guide using a local MongoDB replica set in which there are no auth restrictions.

Do you think it would be better to move this requirement into one of the steps (e.g. "Make sure your user has ReadWrite permissions for this Data Key setup step") or remove it?

File: source/use-cases/sensitive-data-encryption.txt (hunk 2 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: On second thought, since that link is for encryption-at-rest, the references to Atlas might be confusing. Is it ok to copy and paste a couple of the bullet points from that article?

File: source/use-cases/sensitive-data-encryption.txt [hunk 7]
```diff
@@ -42,39 +42,45 @@ Requirements
...
 
 MongoDB Driver Compatible with CSFLE 
   * For a list of drivers that support CSFLE, refer to the
-    `driver compatibility table <https://docs.mongodb.com/manual/core/security-client-side-encryption/#driver-compatibility-table>`_.
+    :manual:`driver compatibility table </core/security-client-side-encryption/#driver-compatibility-table>`.
 
 
 AWS Key Management Service (KMS)
-  * AWS KMS is required for the secure deployment of CSFLE.
+  * The client application will need the following AWS KMS permissions to the
```
Comment (ccho-mongodb) [inline source/use-cases/sensitive-data-encryption.txt]: Updated list, confirmed by Kenn White.

File: source/use-cases/sensitive-data-encryption.txt (hunk 2 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Let me know what you think about the updates.

File: source/use-cases/sensitive-data-encryption.txt (hunk 6 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: How about **File System Permissions**?

File: source/use-cases/sensitive-data-encryption.txt (hunk 3 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: We definitely should include the required permissions since I presume enterprise users will deploy to production at some point which won't use localhost. We should also include it in the steps as an admonition.

A few things about the mongodb roles:

- Can we reframe the bullet points to be more independent?
   - e.g.
     ```rst
     - To create data encryption keys...
     - To encrypt and decrypt data...
     ```
- These bullets seem inconsistent:
   - `readWrite` is a built-in database user role
   - `listCollections` is a privilege/action granted by both the `read` and `readWrite` built-in database user roles
  - Built-in roles apply to entire databases, so you would have a built-in role _on the database that contains the keyvault collection._

File: source/use-cases/sensitive-data-encryption.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: I think this is fine with CSFLE. The context of the guide should make it clear and the link out can clarify any confusion.

File: source/use-cases/sensitive-data-encryption.txt (hunk 2 above)
Comment (nlarew) [reply source/use-cases/sensitive-data-encryption.txt]: These look good to me

File: source/use-cases/sensitive-data-encryption.txt (hunk 3 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Discussed in-person; will remove this section and create a note in the Data Encryption Key section that explains what permissions are needed to create the key only. FLE team still waffling on where best to keep keys secure and standard naming across examples.

File: source/use-cases/sensitive-data-encryption.txt (hunk 6 above)
Comment (ccho-mongodb) [reply source/use-cases/sensitive-data-encryption.txt]: Works for me, thanks.
